ion-schema/src/isl/version_based_isl/mod.rs

```diff
@@ -0,0 +1,396 @@
+    /// For more information: https://amazon-ion.github.io/ion-schema/docs/isl-1-0/spec#type-definitions
+    types: Vec<IslType<V>>,
+    /// Represents all the inline IslImportTypes in this schema file.
+    inline_imported_types: Vec<IslImportType>,
+    /// Represents open content as `Element`s
+    /// Note: This doesn't preserve the information about where the open content is placed in the schema file.
+    /// e.g. This method doesn't differentiate between following schemas with open content:
+    /// ```ion
+    /// $foo
+    /// $bar
+    /// type::{ name: foo, codepoint_length: 1 }
+    /// ```
+    ///
+    /// ```ion
+    /// type::{ name: foo, codepoint_length: 1 }
+    /// $foo
+    /// $bar
+    /// ```
+    open_content: Vec<Element>,
+    phantom: PhantomData<V>,
+}
+
+impl<V: IslVersionTrait> IslSchema<V> {
+    pub fn new<A: AsRef<str>>(
+        id: A,
+        imports: Vec<IslImport>,
+        types: Vec<IslType<V>>,
+        inline_imports: Vec<IslImportType>,
+        open_content: Vec<Element>,
+    ) -> Self {
+        Self {
+            id: id.as_ref().to_owned(),
+            user_reserved_fields: None,
+            imports,
+            types,
+            inline_imported_types: inline_imports,
+            open_content,
+            phantom: Default::default(),
+        }
+    }
+
+    pub fn with_user_reserved_fields(self, user_reserved_fields: UserReservedFields<V>) -> Self {
+        Self {
+            id: self.id,
+            user_reserved_fields: Some(user_reserved_fields),
+            imports: self.imports,
+            types: self.types,
+            inline_imported_types: self.inline_imported_types,
+            open_content: self.open_content,
+            phantom: Default::default(),
+        }
+    }
+
+    pub fn id(&self) -> String {
+        self.id.to_owned()
+    }
+
+    pub fn imports(&self) -> &[IslImport] {
+        &self.imports
+    }
+
+    pub fn types(&self) -> &[IslType<V>] {
+        &self.types
+    }
+
+    pub fn inline_imported_types(&self) -> &[IslImportType] {
+        &self.inline_imported_types
+    }
+
+    /// Provides top level open content for given schema
+    /// For open content defined within type definitions use IslType#open_content()
+    pub fn open_content(&self) -> &Vec<Element> {
+        &self.open_content
+    }
+
+    /// Provide user reserved field defined in the given schema for ISL 2.0,
+    /// Otherwise returns None
+    pub fn user_reserved_fields(&self) -> Option<&UserReservedFields<V>> {
+        self.user_reserved_fields.as_ref()
+    }
+}
+
+/// Represents both named and anonymous [IslType]s and can be converted to a resolved type definition
+/// Named ISL type grammar: `type:: { name: <NAME>, <CONSTRAINT>...}`
+/// Anonymous ISL type grammar: `{ <CONSTRAINT>... }`
+#[derive(Debug, Clone, PartialEq)]
+pub struct IslType<V: IslVersionTrait> {
+    name: Option<String>,
+    constraints: Vec<IslConstraint<V>>,
```

Comment:
suggestion—I recommend using a Bag data type for this, but I am okay with deferring that for now.

Reply:
Just curious, do we already have a `Bag` type implemented somewhere? Or did you have a 3P impl in mind?

Reply:
We don't have a bag type. There are plenty of 3P bag/multiset types available, but I don't have a particular one in mind. Either way, we need it so that we have proper equality since the order of the constraints is unimportant.

If open content was modeled separately from the constraints, we could use a set instead of a bag since two completely identical constraints are redundant and functionally equivalent to having only one of them.